### PR TITLE
chore: add tests for browser connect and fix for browser close for co…

### PR DESCRIPTION
**Pull request by asambstack**
Added tests for browsertype connect.
Fixes https://github.com/microsoft/playwright-dotnet/issues/1784

Relates to https://github.com/microsoft/playwright-dotnet/pull/1602

### CODE REVIEW

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 1]
```diff
@@ -80,21 +80,54 @@ public void TearDown()
         }
 
         [PlaywrightTest("browsertype-connect.spec.ts", "should be able to reconnect to a browser")]
-        [Ignore("SKIP WIRE")]
-        public void ShouldBeAbleToConnectToBrowserAsync()
+        public async Task ShouldBeAbleToConnectToBrowserAsync()
```
Comment (pavelfeldman) [inline src/Playwright.Tests/BrowserTypeConnectTests.cs:83]: For all these tests, you should take the original ("should be able to reconnect to a browser" in browsertype-connect.spec.ts for this one), and port them here. Your tests don't seem to be checking the same scenarios.

Comment (asambstack) [reply src/Playwright.Tests/BrowserTypeConnectTests.cs:83]: Fixed the tests.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 2]
```diff
@@ -80,21 +80,70 @@ public void TearDown()
...
+            Assert.AreEqual(result.ToString(), "121");
+            await browser2.CloseAsync();
         }
 
         [PlaywrightTest("browsertype-connect.spec.ts", "disconnected event should be emitted when browser is closed or server is closed")]
-        [Ignore("SKIP WIRE")]
-        public void DisconnectedEventShouldBeEmittedWhenBrowserIsClosedOrServerIsClosed()
+        public async Task DisconnectedEventShouldBeEmittedWhenBrowserIsClosedOrServerIsClosed()
+        {
+            var browserOne = await BrowserType.ConnectAsync(_browserServer.WSEndpoint);
```
Comment (pavelfeldman) [inline src/Playwright.Tests/BrowserTypeConnectTests.cs:129]: This test does not seem to be on par with the one in https://github.com/microsoft/playwright/blob/main/tests/browsertype-connect.spec.ts.

We should port all the tests as is to ensure connect is flawless.

**Comment by pavelfeldman** [comment]
In order to expose browserType.connect, we should have all the tests from https://github.com/microsoft/playwright/blob/main/tests/browsertype-connect.spec.ts ported and passing. This is substantial amount of code and reviewing it is time-consuming. You are saying "Fixed the tests.", but I can see that some of the tests are incomplete, while other tests are missing. I'm happy to review it again once the tests are in place.

**Comment by asambstack** [comment]
Hey @pavelfeldman , thanks a lot for your feedback!
I've rewritten the tests and ported most of the specs from https://github.com/microsoft/playwright/blob/main/tests/browsertype-connect.spec.ts. The selector feature isn't available for `ConnectAsync` currently, hence haven't included the test for that. Will add that feature in a future PR for the selector support. Have made sure that the all the tests in dotnet are testing the same flow as that of NodeJS.
Looking forward to your review!

**Comment by asambstack** [comment]
Hey @pavelfeldman 
Bumping this thread. Did you get a chance to review the tests?

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 3]
```diff
@@ -80,71 +82,396 @@ public void TearDown()
...
+        }
+
+        [PlaywrightTest("browsertype-connect.spec.ts", "should support slowmo option")]
+        public async Task ShouldSupportSlowMo()
         {
+            var browser = await BrowserType.ConnectAsync(_browserServer.WSEndpoint, new BrowserTypeConnectOptions { SlowMo = 200 });
+            var context = await browser.NewContextAsync();
+            var time1 = DateTime.Now;
+            var page = await context.NewPageAsync();
+            await page.GotoAsync(Server.EmptyPage);
```
Comment (pavelfeldman) [inline src/Playwright.Tests/BrowserTypeConnectTests.cs:143]: I don't see these lines in the original test.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 4]
```diff
@@ -80,71 +82,396 @@ public void TearDown()
...
-        public void ShouldHandleExceptionsDuringConnect()
+        [PlaywrightTest("browsertype-connect.spec.ts", "disconnected event should have browser as argument")]
+        public async Task DisconnectedEventShouldHaveBrowserAsArguments()
         {
+            var browser = await BrowserType.ConnectAsync(_browserServer.WSEndpoint);
+            IBrowser disconneced = null;
+            browser.Disconnected += (_, e) => disconneced = e;
+            await browser.CloseAsync();
+            Assert.AreEqual(browser, disconneced);
         }
```
Comment (pavelfeldman) [inline src/Playwright.Tests/BrowserTypeConnectTests.cs:181]: "should handle exceptions during connect" test is missing

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 5]
```diff
@@ -80,71 +82,396 @@ public void TearDown()
...
-        [Ignore("SKIP WIRE")]
-        public void ShouldRejectNavigationWhenBrowserCloses()
+        public async Task ShouldRejectNavigationWhenBrowserCloses()
         {
+            var browser = await BrowserType.ConnectAsync(_browserServer.WSEndpoint);
+            var page = await browser.NewPageAsync();
+            await browser.CloseAsync();
+
+            Assert.AreEqual(browser.IsConnected, false);
+            var exception = await PlaywrightAssert.ThrowsAsync<PlaywrightException>(async () => await page.GotoAsync(Server.EmptyPage));
```
Comment (pavelfeldman) [inline src/Playwright.Tests/BrowserTypeConnectTests.cs:222]: This is not what the original test does.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 6]
```diff
@@ -80,71 +82,396 @@ public void TearDown()
...
+            var task = TaskUtils.WhenAll(page.WaitForRequestAsync(Server.EmptyPage), page.WaitForResponseAsync(Server.EmptyPage), browser.CloseAsync());
+            try
+            {
+                await task;
+            }
+            catch (Exception)
+            {
+                foreach (var exception in task.Exception.InnerExceptions)
+                {
+                    Console.WriteLine(exception.Message);
```
Comment (pavelfeldman) [inline src/Playwright.Tests/BrowserTypeConnectTests.cs:271]: Remove debug logging

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 7]
```diff
@@ -80,71 +82,396 @@ public void TearDown()
...
+            var context = await browser.NewContextAsync();
+            var page = await context.NewPageAsync();
+            var task = TaskUtils.WhenAll(page.WaitForRequestAsync(Server.EmptyPage), page.WaitForResponseAsync(Server.EmptyPage), browser.CloseAsync());
+            try
+            {
+                await task;
+            }
+            catch (Exception)
+            {
+                foreach (var exception in task.Exception.InnerExceptions)
```
Comment (pavelfeldman) [inline src/Playwright.Tests/BrowserTypeConnectTests.cs:269]: You test will pass (false positive) if there are no exceptions.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 8]
```diff
@@ -80,71 +82,396 @@ public void TearDown()
...
+                {
+                    Console.WriteLine(exception.Message);
+                    StringAssert.Contains("Page closed", exception.Message);
+                }
+            }
+        }
+
```
Comment (pavelfeldman) [inline src/Playwright.Tests/BrowserTypeConnectTests.cs:276]: "should respect selectors" test is missing.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 9]
```diff
@@ -80,71 +82,396 @@ public void TearDown()
...
+        [PlaywrightTest("browsertype-connect.spec.ts", "should not throw on close after disconnect")]
+        public async Task ShouldNotThrowOnCloseAfterDisconnect()
+        {
+            var browser = await BrowserType.ConnectAsync(_browserServer.WSEndpoint);
+            var page = await browser.NewPageAsync();
+
+            _browserServer.Process.Kill();
+            await browser.CloseAsync();
+            while (browser.IsConnected)
```
Comment (pavelfeldman) [inline src/Playwright.Tests/BrowserTypeConnectTests.cs:285]: style: here and below, can this be put in a form of a task as in the original test?

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 10]
```diff
@@ -80,71 +82,396 @@ public void TearDown()
...
+                page.ClickAsync("a"));
+
+            using var tmpDir = new TempDirectory();
+            string userPath = Path.Combine(tmpDir.Path, "download.txt");
+            var download = downloadTask.Result;
+            await download.DeleteAsync();
+            var exception = await PlaywrightAssert.ThrowsAsync<PlaywrightException>(() => download.SaveAsAsync(userPath));
+            StringAssert.Contains("Target page, context or browser has been closed", exception.Message);
+        }
+
```
Comment (pavelfeldman) [inline src/Playwright.Tests/BrowserTypeConnectTests.cs:442]: "should properly disconnect when connection closes from the client side" test is missing
"should properly disconnect when connection closes from the server side" test is missing

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 11]
```diff
@@ -80,71 +82,396 @@ public void TearDown()
...
+            var page = await context.NewPageAsync();
+            await page.GotoAsync(Server.EmptyPage);
+            await context.CloseAsync();
+            await browser.CloseAsync();
+
+
+            Assert.That(harPath, Does.Exist);
+            var logString = System.IO.File.ReadAllText(harPath);
+            StringAssert.Contains(Server.EmptyPage, logString);
+        }
```
Comment (pavelfeldman) [inline src/Playwright.Tests/BrowserTypeConnectTests.cs:475]: "should record trace with sources" test is missing

File: src/Playwright/Core/BrowserContext.cs [hunk 12]
```diff
@@ -166,6 +168,11 @@ public async Task CloseAsync()
         {
             try
             {
+                if (IsClosed)
```
Comment (pavelfeldman) [inline src/Playwright/Core/BrowserContext.cs:171]: Rename this to IsClosedOrClosing, set it from Channel_Closed handler.

File: src/Playwright/Core/BrowserType.cs [hunk 13]
```diff
@@ -149,8 +149,23 @@ public async Task<IBrowser> ConnectAsync(string wsEndpoint, BrowserTypeConnectOp
...
             }
+            var browser = playwright.PreLaunchedBrowser;
+            browser.IsRemote = true;
+            browser.IsConnectedOverWebSocket = true;
+            browser.Disconnected += (_, e) => CloseBrowser(transport, browser);
+            return browser;
+        }
 
-            return playwright.PreLaunchedBrowser;
+        private void CloseBrowser(WebSocketTransport transport, Browser browser)
```
Comment (pavelfeldman) [inline src/Playwright/Core/BrowserType.cs:159]: Rename to be specific to the remote mode.

File: src/Playwright/Core/Frame.cs [hunk 14]
```diff
@@ -195,6 +195,11 @@ public async Task WaitForLoadStateAsync(LoadState? state = default, FrameWaitFor
 
         public async Task<IResponse> WaitForNavigationAsync(FrameWaitForNavigationOptions options = default)
         {
+            if (Page.IsClosed)
```
Comment (pavelfeldman) [inline src/Playwright/Core/Frame.cs:198]: If you go this path, you need to add 300 checks like this in all API methods. The point is that underlying transport error is bubbled every time the call is made after connection is closed.

File: src/Playwright/Core/Page.cs [hunk 15]
```diff
@@ -449,6 +449,10 @@ public async Task CloseAsync(PageCloseOptions options = default)
         {
             try
             {
+                if (IsClosed)
```
Comment (pavelfeldman) [inline src/Playwright/Core/Page.cs:452]: Why is this check here? You can still call CloseAsync twice, even with this check.

**Review by pavelfeldman — COMMENTED** [review]
The tests look much better now! The check in Page for isClosed is not good though.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs (hunk 8 above)
Comment (asambstack) [reply src/Playwright.Tests/BrowserTypeConnectTests.cs:276]: The selectors functionality isn't available in `ConnectAsync` currently. Will add it in a follow up PR.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs [hunk 16]
```diff
@@ -80,71 +82,396 @@ public void TearDown()
...
+        [Skip(SkipAttribute.Targets.Windows)]
+        public async Task ShouldTimeoutInConnectWhileConnecting()
+        {
+            var exception = await PlaywrightAssert.ThrowsAsync<PlaywrightException>(async () => await BrowserType.ConnectAsync(_browserServer.WSEndpoint, new BrowserTypeConnectOptions { Timeout = 1 }));
+            StringAssert.Contains("Timeout 1ms exceeded", exception.Message);
+        }
+
+        [PlaywrightTest("browsertype-connect.spec.ts", "should support slowmo option")]
+        public async Task ShouldSupportSlowMo()
         {
```
[the remark this answers is not in the document]
Comment (asambstack) [reply src/Playwright.Tests/BrowserTypeConnectTests.cs:138]: These tests require us to use a server to connect and record the headers. I wasn't able to use the existing "SimpleServer" for the tests. Will need to figure this out and update the server. Doing it in the follow up PR.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs (hunk 3 above)
Comment (asambstack) [reply src/Playwright.Tests/BrowserTypeConnectTests.cs:143]: Have updated it.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs (hunk 7 above)
Comment (asambstack) [reply src/Playwright.Tests/BrowserTypeConnectTests.cs:269]: Have fixed this test

File: src/Playwright.Tests/BrowserTypeConnectTests.cs (hunk 6 above)
Comment (asambstack) [reply src/Playwright.Tests/BrowserTypeConnectTests.cs:271]: Removed the logging.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs (hunk 9 above)
Comment (asambstack) [reply src/Playwright.Tests/BrowserTypeConnectTests.cs:285]: Created a private function to close the client and server.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs (hunk 11 above)
Comment (asambstack) [reply src/Playwright.Tests/BrowserTypeConnectTests.cs:475]: Added this test.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs (hunk 10 above)
Comment (asambstack) [reply src/Playwright.Tests/BrowserTypeConnectTests.cs:442]: As discussed above, need to update the SimpleServer for these tests. Will create a separate PR for modifying "SimpleServer.cs" and writing the tests.

File: src/Playwright/Core/BrowserContext.cs (hunk 12 above)
Comment (asambstack) [reply src/Playwright/Core/BrowserContext.cs:171]: Done

File: src/Playwright/Core/Page.cs (hunk 15 above)
Comment (asambstack) [reply src/Playwright/Core/Page.cs:452]: If this check isn't put there, it'll throw when calling CloseAsync. Doing something similar to Playwright Java - https://github.com/microsoft/playwright-java/blob/main/playwright/src/main/java/com/microsoft/playwright/impl/PageImpl.java#L542

File: src/Playwright/Core/BrowserType.cs (hunk 13 above)
Comment (asambstack) [reply src/Playwright/Core/BrowserType.cs:159]: Renamed it.

File: src/Playwright.Tests/BrowserTypeConnectTests.cs (hunk 5 above)
Comment (asambstack) [reply src/Playwright.Tests/BrowserTypeConnectTests.cs]: Have changed it accordingly

File: src/Playwright.Tests/BrowserTypeConnectTests.cs (hunk 4 above)
Comment (asambstack) [reply src/Playwright.Tests/BrowserTypeConnectTests.cs:181]: This seems to be an implementational detail for NodeJS. Test isn't present on playwright java as well - https://github.com/microsoft/playwright-java/blob/main/playwright/src/test/java/com/microsoft/playwright/TestBrowserTypeConnect.java#L211

**Comment by asambstack** [comment]
Hey @pavelfeldman 
Did you get a chance to review these tests?

**Comment by asambstack** [comment]
Hey @pavelfeldman 
Gentle bump

**Comment by mxschmitt** [comment]
Merged and I'll follow up with #1908.

**Comment by asambstack** [comment]
Thanks @mxschmitt !
So are we releasing support for connect in the next release?

**Pull request merged**